### ISSUE TITLE
Add max-norm regularization

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#3898]
http://jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf

Idea: constrain the L2 norm of the weights into a each hidden unit to  be less than or equal to a constant C.

**Comment by maxpumperla** [comment]
@AlexDBlack maybe we can have a module similar to keras' `constraints`, see https://keras.io/constraints/

**Comment by mbkennel** [comment]
There are multiple potential constraints which can be useful, constraining the average squared weight of all weights of a layer, i.e. one constraint for the layer, or a number of constraints equal to the number of the number of hiddens (could be outputs potentially) for that layer.   L2 constraining is very simple too, project to the ball's surface by the obvious scaling.   

Only question is how these may affect the various adaptive gradient algorithms and Nesterov momentum, and where to implement in the appropriate place.  E.g. for momentum, do we consider the "previous step" to be where the unconstrained step would have taken it, or where it ended up with the constraint?  These issues should be explicitly recognized and not silent accidents of implementation.

**Comment by AlexDBlack** [comment]
Implemented here: https://github.com/deeplearning4j/deeplearning4j/pull/3957
